src/format.rs

```diff
@@ -0,0 +1,1028 @@
+                let (hour, minute, second) = (time.hour(), time.minute(), time.second());
+                write!(f, "{:02}:{:02}:{:02}", hour, minute, second)
+            }
+        }
+    }
+}
+
+pub(crate) struct TimeFormatter<'t, 'f, T> {
+    time: &'t T,
+    format: &'f [u8],
+}
+
+impl<'t, 'f, T: Time> TimeFormatter<'t, 'f, T> {
+    pub(crate) fn new<F: AsRef<[u8]> + ?Sized>(time: &'t T, format: &'f F) -> Self {
+        Self {
+            time,
+            format: format.as_ref(),
+        }
+    }
+
+    pub(crate) fn fmt(&self, buf: &mut dyn Write) -> Result<(), Error> {
+        // Use a size limiter to limit the maximum size of the resulting formatted string
+        let size_limit = self.format.len().saturating_mul(512 * 1024).max(1024);
+        let mut f = SizeLimiter::new(buf, size_limit);
+
+        let mut cursor = Cursor::new(self.format);
+
+        loop {
+            f.write_all(cursor.read_until(|&x| x == b'%'))?;
+
+            let remaining_before = cursor.remaining();
+
+            // Read the '%' character
+            if cursor.next().is_none() {
+                break;
+            }
+
+            match Self::parse_spec(&mut cursor)? {
+                Some(piece) => piece.fmt(&mut f, self.time)?,
+                None => {
+                    // No valid format specifier was found
+                    let remaining_after = cursor.remaining();
+                    let text = &remaining_before[..remaining_before.len() - remaining_after.len()];
+                    f.write_all(text)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn parse_spec(cursor: &mut Cursor<'_>) -> Result<Option<Piece>, Error> {
+        // Parse flags
+        let mut padding = Padding::default();
+        let mut flags = Flags::empty();
+
+        loop {
+            // The left padding overrides the other padding options for most cases.
+            // It is also used for the hour sign in the %z specifier.
+            //
+            // Similary, the change case flag overrides the upper case flag, except
+            // when using combination specifiers (%c, %D, %x, %F, %v, %r, %R, %T, %X).
+            match cursor.remaining().first() {
+                Some(&b'-') => {
+                    padding = Padding::Left;
+                    flags.insert(Flags::LEFT_PADDING);
+                }
+                Some(&b'_') => padding = Padding::Spaces,
+                Some(&b'0') => padding = Padding::Zeros,
+                Some(&b'^') => flags.insert(Flags::UPPER_CASE),
+                Some(&b'#') => flags.insert(Flags::CHANGE_CASE),
+                _ => break,
+            }
+            cursor.next();
+        }
+
+        // Parse width
+        let width_digits = str::from_utf8(cursor.read_while(u8::is_ascii_digit))
+            .expect("reading ASCII digits should yield a valid UTF-8 slice");
+
+        let width = match width_digits.parse::<usize>() {
```

Comment:
After digging some more into the `INT_MAX` C constant, I think it's defined in terms of the C `int` type. The `int` type is signed, so if we are to use `usize` here, I think `isize` is more correct.

If the `std` feature is enabled, I'd love us to use [`std::os::raw::c_int`](https://doc.rust-lang.org/stable/std/os/raw/type.c_int.html). In the absence of the `std` feature, I'd feel more comfortable if we defaulted to `i32`.

Maybe we need a `cfg`'d type alias?

Reply:
It looks like the `c_int` type alias will be available in `core` in Rust 1.64.0. The feature was recently stabilized:

- https://github.com/rust-lang/rust/pull/98315